apps/cmp.c

```diff
@@ -2697,13 +2697,13 @@ int cmp_main(int argc, char **argv)
     ret = 0;
 
     if (opt_batch) {
-        UI_METHOD *ui_fallback_method;
 #ifndef OPENSSL_NO_UI_CONSOLE
-        ui_fallback_method = UI_OpenSSL();
-#else
-        ui_fallback_method = (UI_METHOD *)UI_null();
+        /*        static UI_METHOD ui_null = {"NULL UI",
+                                    NULL, NULL, NULL, NULL, NULL, NULL};
+        UI_METHOD *ui_fallback_method = (UI_METHOD *)UI_null();
+        */
+        UI_method_set_reader(UI_OpenSSL(), NULL);
```

Comment:
Should we really be "setting" this OpenSSL global method? Shouldn't it be declared as const to prevent this?
I was expecting a solution that copied the UI_METHOD instead. In fact this code looks very much like this existing code where we do make a copy:

https://github.com/openssl/openssl/blob/9524a3089c0d4c3ae553e86f877eef9d6b192b20/apps/lib/apps_ui.c#L115-L129

Reply:
> Shouldn't it be declared as const to prevent this?

Although changing the return value of `UI_OpenSSL()` itself is probably an API break...so maybe not.

Reply:
The code adapting the UI method `cmp.c` when the `-batch` option is given had indeed been borrowed from an earlier version of the above quoted code in `apps_ui.c`.

The goal is to disable any interactive app input, and this so far has been achieved by changing the UI method globally. To admit, this was a pretty invasive way of doing that. 
I've just had a closer look and found a more elegant way, namely to modify only the app-specific UI method. This required removing `const` from the declaration of `UI_METHOD *get_ui_method(void)`, which I believe is adequate.

Reply:
I'm still slightly confused....where is the ui method actually *used* by the cmp app?

Reply:
Sigh, it turns out that the more elegant version does not work because `load_key_certs_crls()` relies on the UI method returned by `get_ui_method()`, so changed the fix back to adapting `UI_OpenSSL()` instead.

Reply:
> where is the ui method actually used by the cmp app?

The UI methods are used by all apps using `load_key()` and the like, which rely on `load_key_certs_crls()`.

Reply:
I suppose the clean fix would be to to add set_batch_ui_method() which would set the NULL reader. in the ui_method returned by get_ui_method?

Reply:
As I tried to explain above removing the reader method from the apps' `ui_method` (regardless which way this is done) is not the right solution because `load_key_certs_crls()` unfortunately uses this UI method for passing further down the `pass` argument. So any given password is lost this way :-(

Reply:
Ah, now I get it. But you could then set a flag to just return null and not call the reader from ui_fallback_method in case the password is not set.

Reply:
The mentioned problem is in case a password must be used, not when it is not set).

I suggest doing any further discussion on this later 
because getting this fix merged is urgent to fix CI test failures
and I have other very urgent things to do today.